### PR TITLE
Keep left non-matches for nearest neighbors crossmatch

### LINKED ISSUE #151: More options for nearest neighbor selection for KdTreeCrossmatch
Currently, we have two ways to control neighbor selection with the KdTreeCrossmatch algorithm: the maximum number of neighbors to select `n_neighbors` and the maximum radius `radius_arcsec`. However, it would be beneficial to add more flexibility:

- [ ] Set a minimum radius and/or skip the first nth matches. This would be very useful for catalog self-matching tasks.
- [ ] Support a "left" join in addition to the "inner" join we have now, so that rows from the left table that do not match are kept in the result.

I propose that we keep this issue open for discussions while creating separate issues for individual tasks.

### PR DESCRIPTION
**Pull request by camposandro**
Implements the "left" crossmatching strategy that allows to keep non-matches for the left catalog. Closes #151.

### CODE REVIEW

**Comment by camposandro** [comment]
@smcguire-cmu I opened issue https://github.com/astronomy-commons/lsdb/issues/273 to tackle the inconsistency in column data types. Please let me know if you think I should go for it before we actually merge this!

File: src/lsdb/core/crossmatch/abstract_crossmatch_algorithm.py [hunk 1]
```diff
@@ -59,7 +60,7 @@ def __init__(
         self.suffixes = suffixes
 
     @abstractmethod
-    def crossmatch(self) -> pd.DataFrame:
+    def crossmatch(self, how: PixelAlignmentType) -> pd.DataFrame:
```
Comment (smcguire-cmu) [inline src/lsdb/core/crossmatch/abstract_crossmatch_algorithm.py:63]: default inner

File: src/lsdb/dask/merge_catalog_functions.py [hunk 2]
```diff
@@ -170,17 +182,19 @@ def construct_catalog_args(
...
         alignment (PixelAlignment): the PixelAlignment to get pixels from
 
     Returns:
-        a tuple of (primary_pixels, join_pixels) with lists of HealpixPixel objects
+        a tuple of (primary_pixels, join_pixels, aligned_pixels) with lists of HealpixPixel objects
     """
     pixel_mapping = alignment.pixel_mapping
-    make_pixel = np.vectorize(HealpixPixel)
+    make_pixel = np.vectorize(
+        lambda o, p: HealpixPixel(o, p) if o is not None and p is not None else INVALID_PIXEL
```
Comment (smcguire-cmu) [inline src/lsdb/dask/merge_catalog_functions.py:196]: Change`align_catalog_to_partitions` to check for none and change this to none not invalid pixel

File: src/lsdb/dask/merge_catalog_functions.py [hunk 3]
```diff
@@ -273,3 +291,14 @@ def align_catalog_to_partitions(
...
     partitions = get_partition(pixels)
     return list(partitions)
+
+
+def create_no_coverage_crossmatch_df(
+    left_df: pd.DataFrame, suffix: str, meta_df: pd.DataFrame
+) -> pd.DataFrame:
+    """Creates the crossmatch DataFrame for partitions in the left catalog that have no coverage
+    in the right catalog."""
+    no_coverage_df = meta_df.copy(deep=False)
```
Comment (smcguire-cmu) [inline src/lsdb/dask/merge_catalog_functions.py:301]: Can you make the right columns type object with None not float with nan?

File: src/lsdb/dask/merge_catalog_functions.py [hunk 4]
```diff
@@ -273,3 +291,14 @@ def align_catalog_to_partitions(
...
+    for name, _ in left_df.dtypes.items():
+        no_coverage_df[name + suffix] = left_df[name]
```
Comment (smcguire-cmu) [inline src/lsdb/dask/merge_catalog_functions.py:303]:
```suggestion
    no_coverage_df = pd.DataFrame(data=np.full((len(left_df), len(meta_df.columns)), fill_value=None), columns=meta_df.columns, index=left_df.index)
    no_coverage_df[left_df.columns] = left_df
```

Does this work?

Comment (camposandro) [reply src/lsdb/dask/merge_catalog_functions.py]: Yes, I just needed to add the suffixes to left columns. Thanks!

File: src/lsdb/core/crossmatch/kdtree_match.py [hunk 5]
```diff
@@ -82,3 +87,8 @@ def _get_point_coordinates(self) -> Tuple[npt.NDArray[np.float64], npt.NDArray[n
             lat=self.right[self.right_catalog_info.dec_column].to_numpy(),
         )
         return left_xyz, right_xyz
+
+    def _get_extra_columns_df(self, chord_distances: npt.NDArray[np.float64]) -> pd.DataFrame:
```
Comment (delucchi-cmu) [inline src/lsdb/core/crossmatch/kdtree_match.py:91]: Would it be easier to append meta if this is constructed from the start as a NestedFrame?

File: src/lsdb/dask/crossmatch_catalog_data.py [hunk 6]
```diff
@@ -36,26 +37,33 @@ def perform_crossmatch(
     left_pix,
     right_pix,
     right_margin_pix,
+    aligned_pix,
```
Comment (delucchi-cmu) [inline src/lsdb/dask/crossmatch_catalog_data.py:40]: This is a terrifying number of positional arguments. Maybe we should switch to using optional keyword args?

But that shouldn't block this PR.

File: src/lsdb/dask/merge_catalog_functions.py [hunk 7]
```diff
@@ -353,3 +374,14 @@ def align_catalog_to_partitions(
     )
     partitions = get_partition(pixels)
     return list(partitions)
+
+
+def create_no_coverage_crossmatch_df(
+    left_df: pd.DataFrame, suffix: str, meta_df: pd.DataFrame
```
Comment (delucchi-cmu) [inline src/lsdb/dask/merge_catalog_functions.py:380]: Use NestedFrame?

File: tests/lsdb/catalog/test_crossmatch.py [hunk 8]
```diff
@@ -112,6 +113,54 @@ def test_crossmatch_negative_margin(
...
+        assert small_sky_catalog.hc_structure.catalog_info.total_rows == len(xmatched)
+        # Matches are equivalent
+        non_match_cond = xmatched["_dist_arcsec"].isna()
+        matches = xmatched[~non_match_cond]
+        assert np.array_equal(np.sort(xmatch_correct_005["ss_id"]), np.sort(matches["id_small_sky"]))
+        # All matches are within the specified distance
+        assert np.all(matches["_dist_arcsec"] <= 0.005 * 3600)
+        # There is only one row per non-match object
+        non_matches = xmatched[non_match_cond]
+        assert not np.any(non_matches["id_small_sky"].duplicated())
```
Comment (delucchi-cmu) [inline tests/lsdb/catalog/test_crossmatch.py:140]: Maybe also throw in a

assert isinstance(catalog._ddf, nd.NestedFrame)

File: src/lsdb/core/crossmatch/abstract_crossmatch_algorithm.py [hunk 9]
```diff
@@ -206,15 +214,28 @@ def _create_crossmatch_df(
...
+        self, right_idx: npt.NDArray[np.int64], extra_columns: pd.DataFrame
+    ) -> pd.DataFrame:
+        """For the right part we need to decide if we keep the non-matches.
+        If we do, first we get the rows for the valid matches and set their
+        cross-match distance column. Then we fill out the remaining rows for
+        the non-matches with empty (nan) data."""
+        # First, get the points from the right that are matches. They are matches
+        # if the index to the right partition is smaller than the length of the
+        # partition and if it is greater than -1.
+        match_cond = np.where(np.logical_and(right_idx < len(self.right), right_idx > -1))
```
Comment (smcguire-cmu) [inline src/lsdb/core/crossmatch/abstract_crossmatch_algorithm.py:233]: Could probably do with a check for how = inner if the match_cond is all true

File: src/lsdb/dask/crossmatch_catalog_data.py (hunk 6 above)
Comment (smcguire-cmu) [inline src/lsdb/dask/crossmatch_catalog_data.py:40]: Why do we need the `aligned_pix` now?

**Comment by camposandro** [comment]
The code base has changed a lot since this code was developed. I'm closing it but I'm keeping the changes of this branch around so that I can open a fresh PR when the renaming is concluded.

**Pull request closed without merging**